### ISSUE TITLE
Trimming fractional seconds in `date-time` and `http-date` protocol tests

### ISSUE THREAD

**Issue by hlbarber** [awslabs/smithy#1607]
The following `@httpResponse` test asserts that, for `restJson1`, the [`http-date` serialization](https://smithy.io/2.0/spec/protocol-traits.html#timestampformat-trait) should remove the trailing fractional seconds.

https://github.com/awslabs/smithy/blob/15b4374f38b5a767b2d1850f8d1b3be105d3cce8/smithy-aws-protocol-tests/model/restJson1/json-structs.smithy#L583-L599

In contrast, the following protocol test asserts that, for `awsJson1_0`, the date-time serialization should __not__ remove the trailing fractional seconds

https://github.com/awslabs/smithy/blob/15b4374f38b5a767b2d1850f8d1b3be105d3cce8/smithy-aws-protocol-tests/model/awsJson1_1/kitchen-sink.smithy#L762-L773

Is this protocol specific behavior? Should the serialization trim the trailing zero fractional seconds for `restJson1` but not `awsJson1_0`?

The same question exists for the ISO 8601 timestamps.

https://github.com/awslabs/smithy/blob/15b4374f38b5a767b2d1850f8d1b3be105d3cce8/smithy-aws-protocol-tests/model/restJson1/json-structs.smithy#L532-L548

https://github.com/awslabs/smithy/blob/15b4374f38b5a767b2d1850f8d1b3be105d3cce8/smithy-aws-protocol-tests/model/awsJson1_1/kitchen-sink.smithy#L750-L761

**Comment by srchase** [comment]
I don't think that there is any conflict in these tests between the behavior for `restJson1` and `awsJson1_1`.

Currently, all of the tests with fractional seconds use `000`, which gets serialized to a whole second.

The `http-date` documentation states the following:
```
An HTTP date as defined by the IMF-fixdate production in [RFC 7231#section-7.1.1.1](https://tools.ietf.org/html/rfc7231.html#section-7.1.1.1) (for example, Tue, 29 Apr 2014 18:30:38 GMT). Note that in addition to the IMF-fixdate format specified in the RFC, implementations MUST also support optional fractional seconds (for example, Sun, 02 Jan 2000 20:34:56.000 GMT).
```

I've tagged this for needing additional protocol tests that cover non-000 fractional seconds to enforce that they are successfully parsed.

**Comment by srchase** [comment]
Yes. I've updated the existing protocol tests to not have any fractional seconds. The tests with `.000` have been updated.

All the fractional second tests have been made "client-only" to reflect that clients should be able to handle fractional-seconds, but that servers are expected to be consistent and not include `.000`.

**Comment by kstich** [comment]
PR was shipped as part of 1.28.1.
